sdks/java/extensions/kafka-factories/src/main/java/org/apache/beam/sdk/extensions/kafka/factories/FileAwareFactoryFn.java

```diff
@@ -0,0 +1,260 @@
+
+  private final String factoryType;
+  private static final Logger LOG = LoggerFactory.getLogger(FileAwareFactoryFn.class);
+
+  public FileAwareFactoryFn(String factoryType) {
+    Preconditions.checkNotNull(factoryType);
+    this.factoryType = factoryType;
+  }
+
+  protected abstract T createObject(Map<String, Object> config);
+
+  @Override
+  public T apply(Map<String, Object> config) {
+    if (config == null) {
+      return createObject(config);
+    }
+
+    Map<String, Object> processedConfig = new HashMap<>(config);
+
+    String key = "";
+    Object value = null;
+    try {
+      downloadAndProcessExtraFiles();
+
+      for (Map.Entry<String, Object> e : config.entrySet()) {
+        try {
+          key = e.getKey();
+          value = e.getValue();
+          if (value instanceof String) {
+            String originalValue = (String) value;
+            Matcher matcher = PATH_PATTERN.matcher(originalValue);
+            StringBuffer sb = new StringBuffer();
+
+            while (matcher.find()) {
+              String externalPath = matcher.group(1);
+              String secretValue = matcher.group(2);
+              String secretFile = matcher.group(3);
+
+              if (externalPath != null) {
+                try {
+                  String tmpPath = replacePathWithLocal(externalPath);
+                  String localPath = downloadExternalFile(externalPath, tmpPath);
+                  matcher.appendReplacement(sb, Matcher.quoteReplacement(localPath));
+                  LOG.info("Downloaded {} to {}", externalPath, localPath);
+                } catch (IOException io) {
+                  throw new IOException("Failed to download file : " + externalPath, io);
+                }
+              } else if (secretValue != null) {
+                try {
+                  String secretId = secretValue.substring(SECRET_VALUE_PREFIX.length());
+                  String processedSecret =
+                      processSecret(originalValue, secretId, getSecretWithCache(secretId));
+
+                  matcher.appendReplacement(sb, Matcher.quoteReplacement(processedSecret));
+                } catch (IllegalArgumentException ia) {
+                  throw new IllegalArgumentException("Failed to get secret.", ia);
+                }
+              } else if (secretFile != null) {
+                throw new UnsupportedOperationException("Not yet implemented.");
+              }
+            }
+            matcher.appendTail(sb);
+            String processedValue = sb.toString();
+            processedConfig.put(key, processedValue);
+          }
+        } catch (IOException ex) {
+          throw new RuntimeException(
+              "Failed trying to process value " + value + " for key " + key + ".", ex);
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Failed trying to process extra files.", e);
+    }
+
+    LOG.info("ProcessedConfig: {}", processedConfig);
```

Comment:
Doesn't this processed config contain secrets? If so, we probably don't want to log it

Reply:
Can you check the rest of the logs too, just in case?